REST API

yesjinu opened the pull request.

yesjinu committed:
chore: install test dependencies

yesjinu committed:
feat: implement GET tasks/ API handlers

yesjinu committed:
chore: ignore package-lock.json for better code review

yesjinu committed:
chore: remove unused function & adjust spacing

yesjinu committed:
feat: add skeleton codes for post, put, delete api handler

johngrib commented on app/build.gradle:
```diff
@@ -17,6 +17,8 @@ repositories {
 }
 
 dependencies {
+    implementation 'org.jetbrains:annotations:20.1.0'
```
이렇게 새로운 dependency를 추가할 때에는 이것이 무엇이고 어떤 기능이 있고, 어떤 장점이 있어서 선택했는지를 설명해주면 좋겠죠. 하지만 그렇다고 주석에 남기기도 애매하죠? 이럴 땐 자기 자신의 PR에 스스로 코드 리뷰를 남기는 것처럼 (제가 지금 작성하고 있는 것처럼) 소개하는 글을 작성해 봅시다.

yesjinu replied:
`org.jetbrains:annotations`는 메소드 파라미터에 `@NotNull` annotation을 사용하기 위해 추가한 dependency입니다. 

이 패키지에는 `@NotNull` 이외에도 `@Nullable`, `@Nls`, `@NonNls`, `@TestOnly`, `@Contract` 등의 annotation이 포함되어 있는데, 다음과 같은 기능을 제공해 코드 작성 중 발생하는 실수를 예방하거나, localization tool, IDE에 정보를 전달합니다.

| Annotations | Description                                                                                                                 |
|-------------|-----------------------------------------------------------------------------------------------------------------------------|
| `@Nls`      | String에 annotated. 로컬라이징이 필요함                                                                                     |
| `@NonNls`   | String에 annotated. 유저에게 보여지지 않고, 로컬라이징도 필요하지 않은 경우. Annotation이 붙어있으면 로컬라이징 툴이 스킵함 |
| `@TestOnly` | Method, Constructor에 annotated. 테스트에만 호출되는 메소드/생성자임을 나타냄                                               |
| `@Contract` | Method에 annotated. 메소드가 만족해야하는 몇 가지 룰(컨트랙트)을 지정할 수 있고, 이를 어기면 IDE에서 경고                   |
| `@Nullable` | Variable, parameter, return value에 annotated. 이름 그대로 Nullable임을 표시                                                |
| `@NonNull`  | Variable, parameter, return value에 annotated. 이름 그대로 반드시 Null이 아니어야 함을 표시. 어길 시 경고 표시              |

Annotation이 달린 variable, method는 IDE가 제공하는 코드 정적 분석기에 의해 컴파일 전 검사되고, `detected problem`으로 하이라이트 되어 표기된다. 이는 코드를 실행하기 전에 미리 문제가 발생할 수 있는 부분을 알려주어 프로그래머의 실수로 발생하는 에러를 줄여준다.

![](https://resources.jetbrains.com/help/img/idea/2022.1/nullability.animated.gif)

[Jetbrains annotation](https://www.jetbrains.com/help/idea/annotating-source-code.html#bundled-annotationsl)

johngrib replied:
좋습니다! 여기에 추가해야 할 설명이 하나 더 있어 보이는데요, 이런 애노테이션들을 사용하게 되었을 때 어떤 장점이 있는지도 써주세요. 애노테이션에서 요구하는 방식의 사용을 어기게 되면 패널티가 있는지 같은 것들을 쓸 수 있겠죠. 가령 컴파일이 안된다던가... 실행 시작시 경고가 줄줄이 뜬다던가... IDE에서 경고를 띄워 보여준다던가...

yesjinu replied:
설명을 조금 추가해보았습니다 🙏

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,163 @@
+package com.codesoom.assignment;
```
패키지 이름 `assignment`는 어떤 뜻인가요?

yesjinu replied:
과제에서 제공된 코드 베이스에 미리 정해져 있던 패키지 이름입니다. 특별한 의미는 없는 것 같습니다 :)

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,163 @@
+                putMapper(exchange);
+            case "DELETE":
+                deleteMapper(exchange);
+            default:
+                throw new IOException();
```
예외를 던질 때 어떤 상황인지를 설명해주는 메시지를 제공해 보면 좋겠습니다.

yesjinu replied:
아래와 같이 request의 HTTP method가 (서버 요구사항에 근거하여) 다뤄지지 않는다는 내용의 메시지를 추가했습니다.

```java
throw new IOException(String.format("Method %s is not handled", reqMethod));
```

yesjinu replied:
[질문]
`HttpHandler`를 상속받은 클래스는 `handle()`이라는 메소드를 오버라이드해야하는데, 이 메소드는 디폴트로 `IOException()` 메소드 시그니처를 포함하고 있습니다. 위와 같이 '다뤄지지 않는 HTTP method의 request를 받은 상황'의 경우, exception을 던지는 게 맞을지, 400번대 에러를 response로 주는 게 맞을지 고민이 됩니다 🤔

yesjinu replied:
[Discord에서 논의 결과](https://discord.com/channels/718694214119981068/954248055870349332/983758710738403389) exception보다는 400번대 error response가 좋다는 의견이 많아, BAD REQUEST response를 내려주는 식으로 처리했습니다 ✅

johngrib replied:
> [질문] `HttpHandler`를 상속받은 클래스는 `handle()`이라는 메소드를 오버라이드해야하는데, 이 메소드는 디폴트로 `IOException()` 메소드 시그니처를 포함하고 있습니다. 위와 같이 '다뤄지지 않는 HTTP method의 request를 받은 상황'의 경우, exception을 던지는 게 맞을지, 400번대 에러를 response로 주는 게 맞을지 고민이 됩니다 🤔

좋은 질문입니다. 이건 정답이 없는 문제이긴 합니다. 예외를 던져도 되고 안 던져도 됩니다. 그래서 이 문제는 좀 다른 영역으로 넘어가게 됩니다. "애플리케이션이 취급하지 않는 URL 경로를 의미하는 문자열이 입력된 것이 예외적인 상황인가?" 라는 질문을 던져보는 거죠. 한편 여기에서 한 번 더 생각해볼 만한 문제는 클라이언트 에러 HTTP 상태를 의미하는 400번대를 핸들링할 때 백엔드 코드에서 예외를 던지고 받고 하는 것이 바람직한가? 가 있겠죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,163 @@
+        String reqPath = exchange.getRequestURI().getPath();
+
+        if (isReqGettingOneTask(reqPath)) {
+            Long id = Long.parseLong(reqPath.split("/")[2]); // Integer.parseInt() vs. Integer.valueOf() 어떤 것이 좋을까요?
```
`parseInt`와 `valueOf`를 비교해 보고 어떤 것을 사용할지 의사결정하는 시간을 가져 봅시다. 가급적이면 글로 적으면 더 좋겠죠. 노션이라던가 개인 블로그라던가 글을 공유할 곳도 많이 있을 겁니다. 두 메소드는 구현부를 열어보기도 간편하니 어렵지 않으실 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,163 @@
+            Long id = Long.parseLong(reqPath.split("/")[2]); // Integer.parseInt() vs. Integer.valueOf() 어떤 것이 좋을까요?
+            Task task = this.taskService.show(id);
+
+            if (task != null) {
+                sendResponse(exchange, 200, toJSON(task));
+            }
+
+            sendResponse(exchange, 404, "Task not found");
+
+        } else if (isReqGettingAllTasks(reqPath)) {
+            List<Task> tasks = this.taskService.showAll();
+
+            if (tasks.size() > 0) {
+                sendResponse(exchange, 200, toJSON(tasks));
+            }
+
+            sendResponse(exchange, 404, "There's no registered tasks");
+
+        } else {
+            sendResponse(exchange, 400, "This request can not be properly handled");
```
`200`, `204` 등이 반복적으로 나타나는군요. 이 중복을 제거하기 위한 방법을 생각해 봅시다.

yesjinu replied:
HttpStatus 클래스를 만들고 내부에 static 변수를 두는 식으로 중복을 제거해보았습니다.

```java
public class HttpStatus {
    public final static Integer OK = 200;
    public final static Integer CREATED = 201;
    public final static Integer NO_CONTENT = 204;
    public final static Integer BAD_REQUEST = 400;
    public final static Integer NOT_FOUND = 404;
}
```

johngrib replied:
이건 class 보다 enum을 사용해 정의해 보도록 합시다. 하는 김에 RFC 2616 같은 문서의 링크도 주석에 추가해주면 좋겠죠. 겸사겸사 문서도 읽고요.

https://datatracker.ietf.org/doc/html/rfc2616

yesjinu replied:
아래와 같이 수정해보았습니다. ✅ 
```java
/**
 * Reference: HTTP/1.1 # Status Code and Reason Phrase
 * https://datatracker.ietf.org/doc/html/rfc2616#section-6.1.1
 */
public enum HttpStatus {
    OK(200),
    CREATED(201),
    NO_CONTENT(204),
    BAD_REQUEST(400),
    NOT_FOUND(404);

    private final Integer statusCode;

    HttpStatus(Integer statusCode) {
        this.statusCode = statusCode;
    }

    public Integer statusCode() { return statusCode; }
}

```

johngrib replied:
좋습니다. 한 가지 코멘트하자면 `Integer`보다 `int`를 쓰는 것이 좀 더 나을 것 같군요.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,163 @@
+        return path.matches("/tasks/?");
+    }
+
+    private Boolean isReqModifyOneTask(String path) {
+        return path.matches("/tasks/\\d+?");
+    }
```
`?`가 없어도 문제가 없을 것 같습니다.
```suggestion
    private Boolean isReqModifyOneTask(String path) {
        return path.matches("/tasks/\\d+");
    }
```

yesjinu replied:
`/tasks/\\d+/?`라고 적는다는걸 슬래쉬를 하나 빠뜨렸군요 😅 request path의 마지막에 슬래쉬가 포함되는 케이스(ex - `localhost:8080/tasks/1/`)가 있을 것 같아서 `/?`를 추가했었는데, 이 부분은 고려하지 않아도 괜찮을까요?

johngrib replied:
이건 이 링크로 제 입장을 대변하도록 하죠. https://johngrib.github.io/wiki/Postel-s-law/

yesjinu replied:
가능한 관대하게 받기 위해 마지막의 `/` 여부까지 고려하는 것으로 처리하는 것이 좋겠군요 ㅎㅎ 👍

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskController.java:
```diff
@@ -0,0 +1,163 @@
+    private final TaskService taskService = new TaskService();
+    
+    public TaskController() {
+        Task task1 = new Task();
+        task1.setId(1L);
+        task1.setTitle("First");
+
+        Task task2 = new Task();
+        task2.setId(2L);
+        task2.setTitle("Second");
+
+        this.taskService.register(task1);
+        this.taskService.register(task2);
```
일종의 연습 코드로군요. 이런 연습 코드가 여기에 있어야 할 지, 삭제하는 것이 좋을지에 대해서도 고민해 봅시다. 아니면 제 3의 방법도 있겠죠.

yesjinu replied:
확실히 이 위치에 그대로 두는 것은 문제가 있어보입니다. 삭제하는 옵션만 생각해보았는데, 제 3의 방법은 어떤 것이 있을지요? 테스트 코드로 분리한달지 하는 방식이 있을까요?

johngrib replied:
저는 테스트 코드를 의도하고 제 3의 방법을 이야기했어요. 혹시 다른 방법이 또 있을까요?

yesjinu replied:
처음 이 코드를 constructor에 넣은 것은 일종의 database seeding 용도였습니다. (Task 조회 API를 테스트해보려면 이미 등록된 task가 있어야 하니까요)

그런 의미에서 모종의 이유로 Database seeding 용 스크립트를 만든다면, 그곳으로 이 코드가 옮겨져도 좋겠다는 생각을 했습니다 ㅎㅎ

johngrib replied:
좋습니다. "용도가 다른" 코드였군요. 이 코드를 어디론가 다른 파일로 옮겨 보도록 합시다. 이름은 일단 아무렇게나 지어 보세요.

yesjinu replied:
`AppTest.java` 파일 안의 `@BeforeAll` 어노테이션이 달린 `setup()` 메소드 안으로 코드를 옮겼습니다. 이후 테스트 로직을 작성할 때 도움이 될 것으로 기대하고 있습니다.
```
    @BeforeAll
    void setup() {
        this.taskService.register("task 1");
        this.taskService.register("task 2");
    }

```

johngrib replied:
좋습니다! 아까운 코드를 굳이 지우지 않아도 되어 기쁘군요.

johngrib reviewed (commented):
반갑습니다.

yesjinu committed:
fix: fix typo in regex

yesjinu committed:
fix: add missing message in exception

yesjinu reviewed (commented):
리뷰 감사합니다 🙏 더 이야기 나누고 싶은 부분에 댓글을 추가했습니다.

yesjinu committed:
refactor: remove magic string and number

yesjinu committed:
chore: remove unused test

yesjinu committed:
feat: implement tasks CRUD API handler

+ 코드 계층 구조 변경. controllers, services, utils, maps 패키지 하위에 각 클래스들을 정리함

yesjinu committed:
refactor: hide http status code in class

yesjinu committed:
chore: remove unused comment

yesjinu committed:
fix: add missed response

yesjinu committed:
fix: 201 CREATED when register succeed

yesjinu committed:
fix: 204 NO CONTENT when delete succeed

yesjinu committed:
fix: change from throwing error to http response

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/PutContoller.java:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    private void handleModifyOneTask(HttpExchange exchange, String path, String body) throws IOException {
```
메소드 네이밍이 좀 아쉽군요. 메소드 이름을 좀 변경해보거나 메소드 시그니처 위에 JavaDoc을 작성해 보도록 합시다. 목표는 이 코드를 처음 보는 사람이 3초 안에 메소드의 기능을 이해하게 만드는 것입니다.

yesjinu replied:
controller 단에 만든 함수이다보니, 이름에서 '특정 로직을 수행하는 느낌'(ex. `modifyTask`)을 주기보다, '서비스 단의 `modify()` 메소드를 호출하게 하는 간접 함수'라는 의미에서 `handleModifyOneTask`라는 이름을 짓게 되었습니다. 그렇게 어떤 역할을 하는 메소드인지 알아보기 어려운 네이밍이 되었네요 ;; 

다시 보니 이 메소드의 이름을 `modifyTask()`로 바꾸는 것이 훨씬 낫겠다는 생각이 들어서 그렇게 변경하였습니다 👍

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,79 @@
+import com.sun.net.httpserver.HttpHandler;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.*;
```
`import *`의 사용을 지양합시다. 어떤 것이 import 되는지 명시해주는 습관을 길러 봅시다. 아마 IDE가 자동으로 하고 있을텐데, 설정을 좀 바꿔야 할 수도 있을 거에요.

yesjinu replied:
'IDE 세팅에 `Use single class import`가 켜져있는데 왜 자동으로 `java.io.*`로 합쳐지지?' 찾다보니, 한 패키지에서 가져오는 import가 5개 이상이면 하나로 합쳐주는 세팅이 디폴트로 설정되어 있었네요 (이미지)

TaskController 클래스의 경우 `IOException`, `InputStream`, `OutputStream`, `InputStreamReader`, `BefferedReader` 5개가 포함되어 `java.io.*`로 합쳐지고 있었구요. 

이렇게 한 패키지에서 여러 클래스를 가져다 쓸 경우에도 import되는 패키지를 명시적으로 적는 것이 좋을지, IDE에서와 같이 '몇 개 이상이면 *로 쓰자'하고 정하는 게 나을지 고민이 됩니다 🤔 

<img width="1094" alt="Screen Shot 2022-06-09 at 7 07 04 PM" src="https://user-images.githubusercontent.com/45530894/172823018-230bf271-589a-4d4f-a617-c3b789ad1245.png">

johngrib replied:
> 이렇게 한 패키지에서 여러 클래스를 가져다 쓸 경우에도 import되는 패키지를 명시적으로 적는 것이 좋을지, IDE에서와 같이 '몇 개 이상이면 *로 쓰자'하고 정하는 게 나을지 고민이 됩니다 🤔

이런 건 보통 팀에서 결정하는 규칙을 따르게 되죠. 우리는 지금 팀이 아니니까 그냥 제가 컨벤션을 만들어 드린 것이라 할 수 있습니다 ㅎㅎ 코드숨에서는 import * 을 사용하지 않고 모두 import 해서 쓰도록 합시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,79 @@
+
+public class TaskController implements HttpHandler {
+
+    private final TaskService taskService = new TaskService();
+    private final GetController getController = new GetController(taskService);
+    private final PostController postController = new PostController(taskService);
+    private final PutContoller putController = new PutContoller(taskService);
+    private final DeleteController deleteController = new DeleteController(taskService);
```
클래스 이름들을 가만히 읽어 봅시다. 이 클래스 이름들을 통해 우리가 알 수 있는 것이 무엇이 있나요?

yesjinu replied:
`TaskService`는 어떤 일을 할 것인지, `OOOController`는 Control한다는 것인지 명확하지가 않군요.. 관습적으로 router는 `contoller` 어미를, 로직은 `Service`라는 어미를 붙여 이름을 지어왔습니다. 

개인적으로는 `TaskService -> TaskManager`, `OOOController -> OOORouter`로 변경하면 의미가 더 명확해질 것 같은데 종립님이 보시기엔 어떤가요?!

```suggestion
    private final TaskManager taskManager = new TaskManager();
    private final GetRouter getRouter = new GetRouter(taskManager);
    private final PostRouter postRouter = new PostRouter(taskManager);
    private final PutRouter putRouter = new PutRouter(taskManager);
    private final DeleteRouter deleteRouter = new DeleteRouter(taskManager);
```

johngrib replied:
Router를 쓰는 것도 괜찮게 느껴지는군요. 다른 언어에서는 이 단어를 쓰는 경우가 있죠. 다만 Controller라는 단어 자체가 갖는 강력한 명성이 있어서 Route 단어를 쓰면 동료들끼리의 의사소통은 Controller를 쓸 때보다 조금 더 귀찮아질 수도 있겠습니다.

TaskService를 TaskManager로 바꾸는 것은 큰 의미가 없는 활동 같아요. Manager가 Service보다 더 이해하기 쉬운 의미를 갖고 있는 것 같지가 않아서에요. 이번주는 일단 이 이름에 대한 고민은 접어두죠. 2주차나 3주차에 한번 더 생각해 볼 기회가 올 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,64 @@
+
+    private final List<Task> registeredTasks = new ArrayList<>();
+
+    public Task show(Long id) {
+        for (Task task: registeredTasks) {
+            if (task.getId().equals(id)) {
+                return task;
+            }
+        }
+
+        return null;
+    }
```
이렇게 for 를 돌면서 id 에 해당하는 Task를 찾는 방식에 문제가 있다면 무엇일까요?

yesjinu replied:
Task의 개수가 많아질수록 탐색 시간 복잡도가 O(n)이 되어 느려진다는 단점이 있을 것 같습니다. `registeredTasks`를 관리하는 자료 구조를 변경한다면 O(logN)이나 O(1)로 단축할 수 있을 것 같네요!

johngrib replied:
좋습니다. 그렇게 수정해 봅시다.

yesjinu replied:
수정했습니다 ✅

johngrib commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,64 @@
+        return null;
+    }
+
+    private Long nextId() {
```
`synchronized`를 고려할 필요도 있을까요? 아니면 전혀 없을까요?

yesjinu replied:
고려할 필요가 있어보입니다. 동시에 여러 스레드가 `nextId()` 메소드에 접근하는 경우, 같은 Id를 가진 task가 여럿 생성될 수 있겠네요. 이 부분에서 `synchronized`를 이용해 모두가 다른 ID를 가지도록 처리할 수 있어보입니다. 

혹은 Task를 저장할 때 `ConcurrentHashMap` 자료구조를 사용해 이미 같은 ID를 가진 task가 있는 경우 새로 ID를 발급받도록 할 수도 있을 것 같습니다. 조금은 비효율적이지만요.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/Mapper.java:
```diff
@@ -0,0 +1,35 @@
+import java.io.OutputStream;
+import java.util.List;
+
+public class Mapper {
```
클래스 이름이 아쉽군요.

yesjinu replied:
`TaskMapper`로 수정해보았습니다. 서로 다른 클래스의 인스턴스를 변환해주는 라이브러리에서 `ModelMapper`, `MapStruct` 등의 이름을 사용하고 있길래 `Mapper` 이름을 차용했습니다. 더 나은 이름이 있을까요? 🤔

johngrib replied:
Task를 취급하는 것 외의 기능이 없다는 점도 고려할 수 있을 거에요. 확장의 가능성이 없다면 Task 또는 Task를 취급하는 클래스의 이너 클래스로 선언해서 쓸 수도 있겠죠.

yesjinu committed:
feat: handle slash at the end of request path (generously 😊)

Reference: https://johngrib.github.io/wiki/Postel-s-law/

yesjinu committed:
fix: change HTTP status code from class to enum

yesjinu committed:
docs: add rfc reference

yesjinu committed:
refactor: refactor contoller private method name to more readable ones

yesjinu committed:
refactor: introduce explicit imports

johngrib reviewed (commented):
코드숨 과정을 진행하는 동안 이 버튼은 가능한 한 직접 누르지 마시고 제가 클릭하게 해 주세요. 한번에 여러 PR을 리뷰하다보면 어떤 건이 끝났고 어떤 건이 안 끝났는지 꽤 헷갈리거든요.

![image](https://user-images.githubusercontent.com/1855714/173100378-83d81e08-8a02-42d7-9dd1-70711179171a.png)

johngrib reviewed (commented):
![image](https://user-images.githubusercontent.com/1855714/173100784-488bf007-b333-4580-8d59-dbb10b0755b6.png)

리퀘스트 요청자가 URL을 잘못 적은 것이 IOException 으로 던질만한 일일까요? 이건 회사마다, 팀마다, 프레임워크마다 접근방법이 조금씩 다를 수 있을 거에요. 그러니 여기엔 정답이 없고, IOException을 던지는 것이 적절한지에 대해 고민해 봐야 하죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/PathParser.java:
```diff
@@ -14,11 +14,11 @@ public static Boolean isReqRegisterOneTask(String path) {
 
     public static Boolean isReqModifyOneTask(String path) {
-        return path.matches("/tasks/\\d+?");
+        return path.matches("/tasks/\\d+/?");
```
`/tasks/011` 과 같은 입력이 들어온다면 어떻게 될까요? `011`을 어떻게 인식하는지에 따라 굉장히 다른 아이디가 될 수도 있습니다.

yesjinu replied:
그렇겠네요 🤔 경우에 따라서 숫자 부분을 따로 파싱해서 별도로 처리하는 로직이 필요하겠네요 (`011`을 `11`과 같게 처리하도록) 

이 코드의 경우 slash 뒷 부분 전체를 파싱해 `Long.parseLong()` 함수를 거쳐 Long으로 변환하고 있기 때문에 문제가 생기지 않고 있었습니다. 
```java
    public static Long parseId(String path) {
        return Long.parseLong(path.split("/")[2]);
    }

```

johngrib replied:
이건 여러 관점을 고려할 수 있어요. 우리가 쓰는 URL은 원격 자원의 식별자를 의미하죠. 그래서 `/`를 구분자로 삼는 계층구조로 작성하는 거고요. 그러므로 `yesjinu.com/tasks/011` 같은 주소는 yesjinu.com 이라는 컴퓨터의 tasks 자료 중에서 아이디가 `011` 인 데이터를 식별합니다. 이때 이 식별자에서 011 이 숫자인지 문자열인지를 명시하지 않는다는 점이 문제인 거죠. 

한편으로는 URL이 코드보다 오래 간다는 점에 대해 진지하게 생각해볼 필요도 있어요. 지금은 우리가 Java를 쓰고 있고 간편하게 `Long.parseLong`을 써서 `011`을 수 `11`로 파싱하고 있습니다. 그런데 만약 레거시가 `011`을 문자열로 파싱하고 있었다면 어땠을까요? 지금 만들고 있는 프로젝트는 레거시를 대체하는 새로운 애플리케이션을 만드는 것이고요. 그렇다면 `011`과 `11`같은 아이디를 논리적으로 구분해서 돌아가고 있던 레거시 클라이언트는 문제가 생기겠죠. (이게 이번 과제의 요구조건은 아니니까 그냥 상상만 해보자는 것입니다)

한편으로는 0 으로 시작하는 숫자는 8진수로 판별하는 경우가 있으니까 레거시가 `011`을 `9`로 파싱하고 있었을 수도 있습니다. 그리고 한 단계 더 나아가서 생각한다면 지금 만들고 있는 프로젝트가 레거시가 됐을 때도 생각할 수 있겠죠. 앗 지금은 011 을 11로 파싱하고 있는데 새로 만드는 애플리케이션에서의 기준은 무엇이 맞을까. 클라이언트는 분명 식별자로 `/tasks/011`을 제공했는데 내가 이걸 `/tasks/11`로 멋대로 바꿔서 생각해도 되는 걸까? 같은 방식으로요. 물론 이런 것들에 대해 정책과 선택이 있게 되겠죠. 011 을 404로 볼 것인지, 11로 받는 관대한 정책을 세울 것인지에 대해서는 어느쪽도 오답이 아니라는 거죠.

yesjinu replied:
그렇네요. 문자열과 달리 숫자는 같은 표현에 대해서 다양한 해석이 가능하다는 점이 어려움을 만드는 포인트가 되는 것 같습니다. 예로 들어 주신 것처럼, `011`이 10진법 `11`, 8진법 `9`,  문자열 `"011"` 등으로 다양하게 해석 가능하다는 점이요.

한편으로는, 이렇게 API path에 숫자가 포함된 사례 이외에도 고려해야할 사항이 워낙 많기 때문에 path 자체에 `/v1/`, `/v2/`처럼 API를 버전으로 분리하는 경우가 많겠구나 하는 생각도 했습니다. 이것은 좋은 정책일까요? 개발 공수나 일정을 생각하면 합리적인 판단이었을 수도 있고, Backward compatibility를 우회해서 해결하려는 시도 같기도 하구요.

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/HttpStatus.java:
```diff
@@ -1,9 +1,17 @@
+        this.statusCode = statusCode;
+    }
+
+    public Integer statusCode() { return statusCode; }
```
가능한 한 이런 식으로 한 줄로 줄이지 않도록 합시다. (코드숨 컨벤션이라고 생각합시다.)

yesjinu replied:
네! IntelliJ가 기본으로 지정해준 포매팅을 그대로 이용했었는데, 조금 더 의식적으로 내가 필요한 스타일대로 수정해야겠습니다. 🙏

yesjinu replied:
IntelliJ setting > Editor > General > Code Folding > Java > One-line methods를 OFF로 변경해 해결했습니다 ✔️

johngrib commented on app/src/main/java/com/codesoom/assignment/utils/HttpStatus.java:
```diff
@@ -1,5 +1,9 @@
 package com.codesoom.assignment.utils;
 
+/**
+ * Reference: HTTP/1.1 # Status Code and Reason Phrase
+ * https://datatracker.ietf.org/doc/html/rfc2616#section-6.1.1
+ */
```
`@see <a href="url">label</a>` 형식을 사용해 보세요. JavaDoc 문서를 빌드했을 때 문서를 읽는 사람이 클릭해서 이동할 수 있게 될 거에요.

yesjinu replied:
```
/**
 * @see <a href="https://datatracker.ietf.org/doc/html/rfc2616#section-6.1.1">HTTP/1.1 # Status Code and Reason Phrase</a>
 */
```
위와 같은 JavaDoc을 추가해보았습니다. 결과는 아래와 같습니다.
<img width="552" alt="Screen Shot 2022-06-12 at 5 49 59 PM" src="https://user-images.githubusercontent.com/45530894/173225251-9cfbb315-e3d6-4e6e-997a-c2026d3eb2c5.png">

johngrib replied:
좋습니다!

yesjinu committed:
refactor: change data structure for Task storage from ArrayList to HashMap

yesjinu committed:
docs: add JavaDoc for the enum

yesjinu committed:
style: change one line expression for method

yesjinu committed:
refactor: rename Mapper to TaskMapper

yesjinu committed:
feat: add database seeding code for test convenience

yesjinu committed:
refactor: rename class name more clearly

- GetController -> GetRouter
- PostController -> PostRouter
- PutController -> PutRouter
- DeleteController -> DeleteRouter
- TaskService -> TaskManager

johngrib reviewed (approved):
한 주간 고생 많으셨습니다!
